File: Palindrome Partitioning II.cpp
class Solution {
public:
    int minCut(string s) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<vector<bool>> part(s.size(), vector<bool>(s.size(),false));
        
        const int len = s.size();
        int i = 0;
        for(i = len-1; i > -1; i--)
        {
            int j = 0;
            for(j = i; j<len; j++)
            {
                part[i][j] = (((j-i<2)||(part[i+1][j-1]))&&(s[i]==s[j]))?true:false;
            }
        }
        
        vector<int> cuts(s.size()+1, INT_MAX);
        cuts[len] = -1;
        for(i = len-1; i > -1; i--)
        {
            int j = 0;
            for(j = i; j<len; j++)
            {
                if(part[i][j])
                    cuts[i] = min(cuts[i], 1+cuts[j+1]);
            }
        }
        
        return cuts[0];
    }
};

File: Palindrome Partitioning II.java
public class Solution {
    public int minCut(String s) {
        // Start typing your Java solution below
        // DO NOT write main() function
        int len = s.length();
        
        boolean [][]palindrome = new boolean[len][len];
        for(int i = len - 1; i > -1; i--)
        {
            for(int j = i; j < len; j++)
            {
                palindrome[i][j] = (((j-i<2)||palindrome[i+1][j-1]) && (s.charAt(i) == s.charAt(j)))? true: false;
            }
        }
        
        int []mincut = new int[len+1];
        mincut[len] = -1; 
        for(int i = len-1 ; i > -1; i-- )
        {
            mincut[i] = len - i - 1;
            for(int j = i; j < len; j++)
            {
                if(palindrome[i][j])
                {
                    mincut[i] =  Math.min(mincut[i], mincut[j+1]+1);
                }
            }
        }
        
        return mincut[0];
    }
    
    
}